File: Oct17/Example40/app/src/main/java/com/varunshankar/example40/MainActivity.java
package com.varunshankar.example40;

import android.content.Context;
import android.graphics.Color;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.TextView;
import android.widget.Toast;

import java.util.Random;

public class MainActivity extends AppCompatActivity {
    private SensorManager mSensorManager;
    private Sensor mAccelerometer;
    public TextView mTvData;
    private final double mThreshold = 0.2;

    //Previous accelerations
    private double last_x, last_y, last_z;

    //current accelerations
    private double now_x, now_y,now_z;
    private boolean mNotFirstTime;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //Get sensor manager
        mSensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);

        //Get the default accelerometer
        mAccelerometer = mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);

        //Get the text view
        mTvData = (TextView) findViewById(R.id.tv_data);

    }

    private SensorEventListener mListener = new SensorEventListener() {


        @Override
        public void onSensorChanged(SensorEvent sensorEvent) {

            //Get the accelerations
            now_x = sensorEvent.values[0];
            now_y = sensorEvent.values[1];
            now_z = sensorEvent.values[2];

            if(mNotFirstTime){
                double dx = Math.abs(last_x - now_x);
                double dy = Math.abs(last_y - now_y);
                double dz = Math.abs(last_z - now_z);

                //Check if the values of acceleration have changed on any pair of axes
                if( (dx > mThreshold && dy > mThreshold) ||
                        (dx > mThreshold && dz > mThreshold)||
                        (dy > mThreshold && dz > mThreshold)){

                    //Change color of text view to some random color
                    Random rnd = new Random();
                    int color = Color.argb(255,rnd.nextInt(256),rnd.nextInt(256),rnd.nextInt(256));
                    mTvData.setBackgroundColor(color);
                }
            }
            last_x = now_x;
            last_y = now_y;
            last_z = now_z;
            mNotFirstTime = true;
        }


        @Override
        public void onAccuracyChanged(Sensor sensor, int i) {

        }
    };

    @Override
    protected void onResume() {
        super.onResume();
        if(mAccelerometer!=null){
            mSensorManager.registerListener(mListener,mAccelerometer,SensorManager.SENSOR_DELAY_NORMAL);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if(mAccelerometer!=null){
            mSensorManager.unregisterListener(mListener);
        }
    }


}

File: Oct17/Example40/app/src/main/java/com/varunshankar/example40/MainActivity.kt
package com.varunshankar.example40

import android.content.Context
import android.graphics.Color
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import java.util.*

class MainActivity : AppCompatActivity() {
    private var mSensorManager: SensorManager? = null
    private var mAccelerometer: Sensor? = null
    var mTvData: TextView? = null
    private val mThreshold = 2.0

    //Previous accelerations
    private var last_x = 0.0
    private var last_y = 0.0
    private var last_z = 0.0

    //current accelerations
    private var now_x = 0.0
    private var now_y = 0.0
    private var now_z = 0.0
    private var mNotFirstTime = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //Get sensor manager
        mSensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager?

        //Get the default accelerometer
        mAccelerometer = mSensorManager!!.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)

        //Get the text view
        mTvData = findViewById(R.id.tv_data)
    }

    private val mListener: SensorEventListener = object : SensorEventListener {
        override fun onSensorChanged(sensorEvent: SensorEvent) {

            //Get the accelerations
            now_x = sensorEvent.values.get(0).toDouble()
            now_y = sensorEvent.values.get(1).toDouble()
            now_z = sensorEvent.values.get(2).toDouble()
            if (mNotFirstTime) {
                val dx = Math.abs(last_x - now_x)
                val dy = Math.abs(last_y - now_y)
                val dz = Math.abs(last_z - now_z)

                //Check if the values of acceleration have changed on any pair of axes
                if ((dx > mThreshold && dy > mThreshold) || (dx > mThreshold && dz > mThreshold) || (dy > mThreshold && dz > mThreshold)) {

                    //Change color of text view to some random color
                    val rnd = Random()
                    val color =
                        Color.argb(255, rnd.nextInt(256), rnd.nextInt(256), rnd.nextInt(256))
                    mTvData!!.setBackgroundColor(color)
                }
            }
            last_x = now_x
            last_y = now_y
            last_z = now_z
            mNotFirstTime = true
        }

        override fun onAccuracyChanged(sensor: Sensor, i: Int) {}
    }

    override fun onResume() {
        super.onResume()
        if (mAccelerometer != null) {
            mSensorManager!!.registerListener(
                mListener,
                mAccelerometer,
                SensorManager.SENSOR_DELAY_NORMAL
            )
        }
    }

    override protected fun onPause() {
        super.onPause()
        if (mAccelerometer != null) {
            mSensorManager!!.unregisterListener(mListener)
        }
    }
}